File: app/schemas/v1/links.py
"""Request/response schemas for tracked links and ROI attribution."""

from datetime import datetime
from urllib.parse import urlsplit

from pydantic import BaseModel, ConfigDict, Field, field_validator


class TrackedLinkCreateRequest(BaseModel):
    project_id: int = Field(ge=1)
    destination_url: str = Field(min_length=1, max_length=2048)
    opportunity_id: int | None = Field(default=None, ge=1)
    reply_draft_id: int | None = Field(default=None, ge=1)

    @field_validator("destination_url")
    @classmethod
    def validate_destination_url(cls, value: str) -> str:
        value = value.strip()
        parts = urlsplit(value)
        if parts.scheme not in ("http", "https") or not parts.netloc:
            raise ValueError("destination_url must be a valid http(s) URL.")
        return value


class TrackedLinkResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    code: str
    destination_url: str
    opportunity_id: int | None = None
    reply_draft_id: int | None = None
    utm_params: dict[str, str] = Field(default_factory=dict)
    created_at: datetime
    short_path: str
    tracked_url: str
    click_count: int = 0


class RoiRollupRow(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    group_by: str
    key: str
    links: int = 0
    clicks: int = 0


class RoiRollupResponse(BaseModel):
    project_id: int
    rows: list[RoiRollupRow] = Field(default_factory=list)

File: web/lib/api/links.ts
import { API_BASE, apiRequest } from "../api";

/** A tracked short link used for reply ROI attribution. */
export type TrackedLink = {
  id: number;
  project_id: number;
  code: string;
  destination_url: string;
  opportunity_id: number | null;
  reply_draft_id: number | null;
  utm_params: Record<string, string>;
  created_at: string;
  short_path: string;
  tracked_url: string;
  click_count: number;
};

export type RoiRollupRow = {
  group_by: "subreddit" | "buying_stage" | string;
  key: string;
  links: number;
  clicks: number;
};

export type RoiRollupResponse = {
  project_id: number;
  rows: RoiRollupRow[];
};

/** The shareable short URL is served by the API host (e.g. https://api.example.com/r/abcd). */
export function shortLinkUrl(link: Pick<TrackedLink, "short_path">): string {
  return `${API_BASE}${link.short_path}`;
}

export async function createTrackedLink(
  token: string,
  data: {
    project_id: number;
    destination_url: string;
    opportunity_id?: number | null;
    reply_draft_id?: number | null;
  },
): Promise<TrackedLink> {
  return apiRequest<TrackedLink>("/v1/links", { method: "POST", body: JSON.stringify(data) }, token);
}

export async function getTrackedLinks(token: string, projectId?: number | null): Promise<TrackedLink[]> {
  const qs = projectId ? `?project_id=${projectId}` : "";
  return apiRequest<TrackedLink[]>(`/v1/links${qs}`, {}, token);
}

export async function getRoiRollup(token: string, projectId?: number | null): Promise<RoiRollupResponse> {
  const qs = projectId ? `?project_id=${projectId}` : "";
  return apiRequest<RoiRollupResponse>(`/v1/analytics/roi${qs}`, {}, token);
}
